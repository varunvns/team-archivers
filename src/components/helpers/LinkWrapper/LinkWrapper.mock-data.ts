const defaultData = {
  className: 'hover:underline',
  field: {
    value: {
      href: 'https://www.horizontaldigital.com',
      text: 'Link with new tab screen reader text',
      linktype: 'external',
      target: '_blank',
    },
  },
  suppressNewTabIcon: false,
  suppressLinkText: false,
  srOnlyText: 'Only screen readers can access this text',
};

export const internalLink = {
  ...defaultData,
  field: {
    value: {
      ...defaultData.field.value,
      target: undefined,
    },
  },
};

export const suppressIconAndText = {
  ...defaultData,
  suppressNewTabIcon: true,
  suppressLinkText: true,
};

export const noContent = {
  field: {
    value: {},
  },
  suppressLinkText: false,
};

export const fieldAsLinkFieldValue = {
  ...defaultData,
  field: {
    href: 'https://www.horizontaldigital.com',
    text: 'Link Text',
    linktype: 'external',
    target: '_blank',
  },
};

export default defaultData;
