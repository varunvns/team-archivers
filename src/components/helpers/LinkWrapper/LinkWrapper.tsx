// Global
import React from 'react';
import { Link, LinkProps, LinkField } from '@sitecore-jss/sitecore-jss-nextjs';

// Lib
import useExperienceEditor from 'lib/use-experience-editor';

/**
 * This component adds some needed accessibility updates to the JSS Link component
 */

export interface LinkWrapperProps extends LinkProps {
  srOnlyText?: string;
  suppressLinkText?: boolean;
  suppressNewTabIcon?: boolean;
}

const INTERNAL_LINK_REGEX = /^\/|^\#/g;

const LinkWrapper = React.forwardRef(
  (
    {
      children,
      field,
      srOnlyText,
      suppressLinkText,
      suppressNewTabIcon,
      ref,
      ...props
    }: LinkWrapperProps,
    _ref
  ): JSX.Element => {
    // Format field as LinkField for consistency
    const asLinkField = !field.value ? { value: { ...field } } : (field as LinkField);
    const text = suppressLinkText ? '' : asLinkField?.value?.text;
    const target = asLinkField?.value?.target;

    const isEE = useExperienceEditor();

    // In experience editor, do not pass any children but retain basic styling so that double components do not appear when using <Link>
    if (isEE) {
      return (
        <Link
          field={asLinkField}
          showLinkTextWithChildrenPresent={false}
          internalLinkMatcher={INTERNAL_LINK_REGEX}
          ref={typeof ref !== 'string' ? ref : null}
          {...props}
        />
      );
    }

    // If no content is present, don't print
    if (!suppressLinkText && !asLinkField.value.text && !asLinkField.value.href) return <></>;

    return (
      <Link
        field={asLinkField}
        internalLinkMatcher={INTERNAL_LINK_REGEX}
        ref={typeof ref !== 'string' ? ref : null}
        showLinkTextWithChildrenPresent={false}
        {...props}
      >
        {text}
        {children}
        {(target === '_blank' || srOnlyText) && (
          <>
            <span className="sr-only">
              {srOnlyText && srOnlyText}
              {/* Preserve a single space character before SR Tab Text */}
              {target === '_blank' && ' (Opens in a new tab)'}
            </span>
            {/* Icon Goes Here */}
            {!suppressNewTabIcon && target === '_blank' && <span></span>}
          </>
        )}
      </Link>
    );
  }
);

LinkWrapper.displayName = 'LinkWrapper';

export default LinkWrapper;
