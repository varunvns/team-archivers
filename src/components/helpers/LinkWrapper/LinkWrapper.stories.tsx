/* eslint-disable  @typescript-eslint/no-explicit-any */
// TODO: Figure out which are the correct types to replace the "any" type definitions in this file.

// Global
import type { Meta, StoryObj } from '@storybook/react';

// Lib
import { expandObj, flattenObj } from 'lib/object-parser';

// Local
import LinkWrapper from './LinkWrapper';
import defaultData from './LinkWrapper.mock-data';

const meta: Meta<typeof LinkWrapper> = {
  title: 'Helpers/A11y/LinkWrapper',
  component: LinkWrapper,
  argTypes: {
    'field.value.linktype': { control: 'select', options: ['external'] },
    'field.value.target': { control: 'select', options: ['_blank', '_self', '_parent', '_top'] },
  } as any,
};

export default meta;

type Story = StoryObj<typeof LinkWrapper>;

export const Default: Story = {
  render: (args) => {
    return <LinkWrapper {...(expandObj({ ...args }) as any)} />;
  },
  args: {
    ...flattenObj(defaultData),
  },
};
