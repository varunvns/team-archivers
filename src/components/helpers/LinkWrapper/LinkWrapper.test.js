// Global
import dynamic from 'next/dynamic';

// Lib
import { renderComponent, snapshot } from 'lib/testing/utils';

// Local
import LinkWrapper from './LinkWrapper';
import defaultData, {
  fieldAsLinkFieldValue,
  noContent,
  suppressIconAndText,
} from './LinkWrapper.mock-data';

// Mock SvgIcon dynamic import
jest.mock('next/dynamic');

const NEW_TAB_TEXT = '(Opens in a new tab)';

it('adds new tab content appropriately', () => {
  snapshot(LinkWrapper, { componentProps: defaultData });
});

it('allows for field to be of type LinkFieldValue', () => {
  snapshot(LinkWrapper, { componentProps: fieldAsLinkFieldValue });
});

it('can suppress text and icons', () => {
  snapshot(LinkWrapper, { componentProps: suppressIconAndText });
});

// it("doesn't change content in Experience Editor", () => {
//   const component = renderComponent(LinkWrapper, {
//     componentProps: defaultData,
//     sitecoreContext: eeSitecoreContext,
//   });
//   // Expect new tab content to not be injected in Experience Editor
//   const newTabText = component.queryByText(NEW_TAB_TEXT);
//   expect(newTabText).toBe(null);
// });

it("doesn't print if no content is present", () => {
  const component = renderComponent(LinkWrapper, { componentProps: noContent });
  expect(component.container.firstChild).toBe(null);
});
