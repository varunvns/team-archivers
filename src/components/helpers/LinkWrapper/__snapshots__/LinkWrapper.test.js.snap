// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds new tab content appropriately 1`] = `
<a
  class="hover:underline"
  href="https://www.horizontaldigital.com"
  rel="noopener noreferrer"
  target="_blank"
>
  Link with new tab screen reader text
  <span
    class="sr-only"
  >
    Only screen readers can access this text
     (Opens in a new tab)
  </span>
  <span />
</a>
`;

exports[`allows for field to be of type LinkFieldValue 1`] = `
<a
  class="hover:underline"
  href="https://www.horizontaldigital.com"
  rel="noopener noreferrer"
  target="_blank"
>
  Link Text
  <span
    class="sr-only"
  >
    Only screen readers can access this text
     (Opens in a new tab)
  </span>
  <span />
</a>
`;

exports[`can suppress text and icons 1`] = `
<a
  class="hover:underline"
  href="https://www.horizontaldigital.com"
  rel="noopener noreferrer"
  target="_blank"
>
  <span
    class="sr-only"
  >
    Only screen readers can access this text
     (Opens in a new tab)
  </span>
</a>
`;
